src/main.rs

```diff
@@ -37,7 +38,40 @@ struct MarketData {
 }
 
 impl MarketData {
-    fn print_holders_map(&self) {
+    fn validate_holders(&self, hash: &String) -> Vec<FileRequest> {
+        // make a map of holders already printed
+        let mut previous_holders: HashMap<&String, &FileRequest> = HashMap::new();
```

Comment:
If this is called every time RegisterFile is run, only 1 duplicate user needs to be considered. We could pass in that user and check for the existing file_request with the matching id and username and remove them. This lets us avoid the hashmap and reduce it to O(1) space.

Reply:
It might be better to remove validate_holders() and put the logic straight into RegisterFile since its only used there

Reply:
But all of the users are stored in a vector, so shouldn't finding a user within that vector should take O(n) anyway?